src/System.cpp

```diff
@@ -6,13 +6,42 @@ using namespace std;
 
 //------------------------------------------------
 // load parameters and functions
-void System::load(Rcpp::NumericVector x, Rcpp::List args_params) {
+void System::load(Rcpp::List args_data, Rcpp::List args_params) {
   
   // data
-  this->x = x;
-  n_loci = x.size();
+  a = rcpp_to_vector_int(args_data["a"]);
+  r = rcpp_to_vector_int(args_data["r"]);
+  n_loci = a.size();
   
-  // allele frequencies
+  // other parameters
   p = rcpp_to_vector_double(args_params["p"]);
+  c = args_params["c"];
   
 }
+
+//------------------------------------------------
+// initialise Beta-binomial likelihood lookup table for each locus
+void System::init_betabinom_lookup() {
+  
+  betabinom_lookup = vector<vector<double>>(n_loci, vector<double>(1001));
+  for (int i = 0; i < n_loci; ++i) {
+    double tmp1 = lgamma(c) - lgamma(a[i] + r[i] + c);
+    for (int j = 0; j < 1001; ++j) {
+      double pi_ = j / double(1000);
+      double tmp2 = lgamma(a[i] + pi_*c) - lgamma(pi_*c);
+      double tmp3 = lgamma(r[i] + (1.0 - pi_)*c) - lgamma((1.0 - pi_)*c);
+      betabinom_lookup[i][j] = exp(tmp1 + tmp2 + tmp3);

```

Comment:
Given the Beta-binomial as defined [here](https://en.wikipedia.org/wiki/Beta-binomial_distribution), this implemetation is missing some data-dependent but pi-independent terms. Namely,

`double tmp0 = lgamma(a[i] + r[i] + 1) / (lgamma(a[i] + 1) + lgamma(r[i] + 1))`

I think it might make sense to include to (i) ensure our final likelihood is unbiased and (ii) make sure the influence of the emission / transition probabilities is balanced.

Going forward we might also try implementing from the quotient of two Beta functions (https://en.cppreference.com/w/cpp/experimental/special_functions/beta) and compare speed.